react/features/calendar-sync/middleware.js

```diff
@@ -0,0 +1,211 @@
+import { MiddlewareRegistry } from '../base/redux';
+import { APP_LINK_SCHEME, parseURIString } from '../base/util';
+
+import { APP_WILL_MOUNT } from '../app';
+
+import { maybeAddNewKnownDomain, updateCalendarEntryList } from './actions';
+import { REFRESH_CALENDAR_ENTRY_LIST } from './actionTypes';
+
+const FETCH_END_DAYS = 10;
+const FETCH_START_DAYS = -1;
+const MAX_LIST_LENGTH = 10;
+const logger = Logger.getLogger(__filename);
+
+MiddlewareRegistry.register(store => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case APP_WILL_MOUNT:
+        _ensureDefaultServer(store);
+        _fetchCalendarEntries(store, false);
+        break;
+    case REFRESH_CALENDAR_ENTRY_LIST:
+        _fetchCalendarEntries(store, true);
+        break;
+    case SET_ROOM:
+        _parseAndAddDomain(store);
+    }
+
+    return result;
+});
+
+/**
+ * Ensures calendar access if possible and resolves the promise if it's granted.
+ *
+ * @private
+ * @param {boolean} promptForPermission - Flag to tell the app if it should
+ * prompt for a calendar permission if it wasn't granted yet.
+ * @returns {Promise}
+ */
+function _ensureCalendarAccess(promptForPermission) {
+    return new Promise((resolve, reject) => {
+        RNCalendarEvents.authorizationStatus()
+            .then(status => {
+                if (status === 'authorized') {
+                    resolve();
+                } else if (promptForPermission) {
+                    RNCalendarEvents.authorizeEventStore()
+                        .then(result => {
+                            if (result === 'authorized') {
+                                resolve();
+                            } else {
+                                reject(result);
+                            }
+                        })
+                        .catch(error => {
+                            reject(error);
+                        });
+                } else {
+                    reject(status);
+                }
+            })
+            .catch(error => {
+                reject(error);
+            });
+    });
+}
+
+/**
+ * Ensures presence of the default server in the known domains list.
+ *
+ * @private
+ * @param {Object} store - The redux store.
+ * @returns {Promise}
+ */
+function _ensureDefaultServer(store) {
+    const state = store.getState();
+    const defaultURL = parseURIString(
+        state['features/app'].app._getDefaultURL()
+    );
+
+    store.dispatch(maybeAddNewKnownDomain(defaultURL.host));
+}
+
+/**
+ * Reads the user's calendar and updates the stored entries if need be.
+ *
+ * @private
+ * @param {Object} store - The redux store.
+ * @param {boolean} promptForPermission - Flag to tell the app if it should
+ * prompt for a calendar permission if it wasn't granted yet.
+ * @returns {void}
+ */
+function _fetchCalendarEntries(store, promptForPermission) {
+    _ensureCalendarAccess(promptForPermission)
+    .then(() => {
```

Comment:
To further this refactor, you could add the catch right after the _ensureCalendarAccess, do the normal .then, and then start chainin.

```
_ensureCalendarAccess()
.catch(/** can't access so error? **/)
.then(() => RNCalendarEvents.fetchAllEvents)
.then(/** parse results **/)
.catch(/** fetch or parse error **)
```

Reply:
Not sure I understand this part, can you please explain?

Reply:
Currently it looks like this
```
_ensurecalendarAccess()
.then(() => {
    fetchAllEvents()
        .then(/ ** process events **/)
        .catch(/** error fetching **/)
})
.catch(/** error accessing **/)
```

If you're in the school of thought of trying to keep promise chains flat, which I am because otherwise they start reminding me of the callback pyramid, one approach for refactoring could be:

```
_ensurecalendarAccess()
.catch(/** error accessing **/)
.then(() => fetchAllEvents())
.then(() => /** process events **/)
.catch(() => /** error fetching or processing **/) // this could also be moved directly after the fetchAllEvents and before the processing of events
```

So you're basically taking the refactoring one step further by reducing the nested promise chaining, which is enabled by moving the error handling for _ensureCalendarAccess closer to the invocation of _ensureCalendarAccess.

Reply:
Wait, that's totally not the equivalent. Disregard all that I've said about this. Must be my ~~dysgraphia~~ ~~aphasia~~ brain not thinking.

Reply:
Of just Friday... I feel the same